nixos/modules/services/web-apps/cryptpad.nix

```diff
@@ -0,0 +1,213 @@
+            "CRYPTPAD_CONFIG=${cryptpadConfigFile}"
+            "HOME=%S/cryptpad"
+          ];
+          ExecStart = "${cfg.package}/bin/cryptpad";
+          PrivateTmp = true;
+          Restart = "always";
+          StateDirectory = "cryptpad";
+          WorkingDirectory = "%S/cryptpad";
+        };
+      };
+    }
+    (mkIf cfg.confinement {
+      systemd.services.cryptpad = {
+        serviceConfig.BindReadOnlyPaths = [
+          cryptpadConfigFile
+          # apparently needs proc for workers management
+          "/proc"
+          "/dev/urandom"
+        ] ++ (if ! cfg.config.blockDailyCheck then [
+          "/etc/resolv.conf"
+          "/etc/hosts"
+        ] else []);
+        confinement = {
+          enable = true;
+          binSh = null;
+          mode = "chroot-only";
+        };
+      };
+    })
+    (mkIf cfg.configureNginx {
+      assertions = [
+        { assertion = cfg.config.httpUnsafeOrigin != "";
+          message = "services.cryptpad.config.httpUnsafeOrigin is required";
+        }
+        { assertion = strings.hasPrefix "https://" cfg.config.httpUnsafeOrigin;
+          message = "services.cryptpad.config.httpUnsafeOrigin must start with https://";
+        }
+        { assertion = isNull cfg.config.httpSafeOrigin || strings.hasPrefix "https://" cfg.config.httpSafeOrigin;
+          message = "services.cryptpad.config.httpSafeOrigin must start with https:// (or be unset)";
+        }
+      ];
+      services.nginx = {
+        enable = true;
+        recommendedTlsSettings = true;
+
+        # FIXME: Check / compare this with [Nginx module configuration in nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/web-servers/nginx/default.nix).
+        # Find out why Cryptpad has this in their documetation. Does this decrease the security of a Cryptpad install
+        # if not used?
+        # diffie-hellman parameters are used to negotiate keys for your session
+        # generate strong parameters using the following command
+        # ssl_dhparam /etc/nginx/dhparam.pem; # openssl dhparam -out /etc/nginx/dhparam.pem 4096
+        #
+        # sslDhparam = null;
+
+        # FIXME: Check / compare this with [Nginx module configuration in nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/web-servers/nginx/default.nix).
+        # Find out why Cryptpad has this in their documentation. Does this decrease the security of a Cryptpad install
+        # if not used?
+        # replace with the IP address of your resolver
+        # resolver 8.8.8.8 8.8.4.4 1.1.1.1 1.0.0.1 9.9.9.9 149.112.112.112 208.67.222.222 208.67.220.220;
+        #
+        # resolver = {};
+
+        virtualHosts = mkMerge [
+          {
+            "${mainDomain}" = {
+              serverAliases = if isNull cfg.config.httpSafeOrigin then [ ] else [ sandboxDomain ];
+              # NOTE: I see no reason not to enable ACME and forcing SSL if you enable Nginx for
```

Comment:
folks might want to use something else than ACME?
My original code had mkdefault here iirc but I got told it's better practice to leave out; I don't have a hard preference but it should be overridable imo

Reply:
IMHO, we shouldn't provide any reverse proxy configuration.

Reply:
I think a web service should do something useful and secure out of the box with a minimal amount of configuration beyond setting `enable = true;`. Most services I looked at for inspiration do this, albeit not setting `configureNginx` default to `true` maybe.
See https://search.nixos.org/options?channel=23.11&show=services.mastodon.configureNginx , https://search.nixos.org/options?channel=23.11&show=services.lemmy.nginx.enable
Or provide an interface to the underlying nginx configuration (I believe this was brought up earlier in this PR):
https://search.nixos.org/options?channel=23.11&show=services.bookstack.nginx

But that's just my opinion and may not be the common approach.

Reply:
Yes, my remark was quite opinionated. I still believe that we shouldn't make assumptions in place of users using the service. For example, I never use Nginx, I always use Caddy... why about providing a configuration for Caddy then? Why Nginx is the default choice here?

Anyway, just my 2 opinionated cents.

Reply:
Because without any proxy configuration Cryptpad won't work out of the box because it relies on SSL and two separate domain names to function securely. And given the nature of Cryptpad, there really isn't any reason to use it without security best practices in place. Nginx because it appears to be the most used, judging from experience and a quick and dirty word count:
```sh
arrakis:~/code/nixpkgs [master]$ grep -roh nginx . | wc -w
    1667
arrakis:~/code/nixpkgs [master]$ grep -roh caddy . | wc -w
     232
arrakis:~/code/nixpkgs [master]$
```
I wouldn't mind having a caddy configuration built in too though! Something like `configureCaddy` in addition to `configureNginx` and have them both default to `false`? With an assertion that you can't enable both at the same time? Maybe you can copy paste a version of your Caddy config into here? I see no harm in having more options for that.
My personal philosophy is to have as few options enabled as necessary to have a working, batteries included experience.